lib/patch_module.ml

```diff
@@ -0,0 +1,28 @@
+let same_obj (a : Obj.t) (b : Obj.t) = a == b
+
+let iteri_fields o ~f =
+  for i = 0 to Obj.size o - 1 do
+    f i (Obj.field o i)
+  done
+
+let find_in_module m_obj x =
+  let exception Found of int in
+  try
+    iteri_fields m_obj ~f:(fun i field ->
+        if same_obj field x then raise (Found i));
+    None
+  with Found i -> Some i
+
+let set_field m ~src ~dst =
+  let m_obj = Obj.repr m in
+  let dst_obj = Obj.repr dst in
+  let src_obj = Obj.repr (src ()) in
+  if same_obj src_obj dst_obj then ()
+  else
+    match find_in_module m_obj src_obj with
+    | None -> invalid_arg "set_field: field not found"
+    | Some i ->
+        Obj.set_field m_obj i dst_obj;
+        let new_value = Obj.repr (src ()) in
+        if not (same_obj new_value dst_obj) then
```

Comment:
The original code did something a bit different - it would set back the field to its original value and try the next field. I believe this was dead code, so we just error out instead

Reply:
also, we can remove this assertion and do the check in `Mdx_top` instead. In that case, we can directly pass `src` as a value and the type of `set_field` becomes the simpler `set_field : _ -> src:'a -> dst:'a -> unit`.

Reply:
Could we return a result type instead? `Mdx_top.patch_env` will be free to propagate or fail on error but at least `set_field` would be cleaner, after having a look at the code it looks reasonable to propagate the result to`Mdx_top.init` up to `Mdx_test.run_exn`